### PR TITLE
Allow Sync From Salesforce (aka Sync To LMS) to send account creation sign-up emails

### PR DESCRIPTION
**Pull request by sadleb**
### Description
After the program launch logistics "Welcome" email goes out from Campaign Monitor (see "Sample Welcome Email" screenshot below), folks who are Fully Confirmed after that still need a link to create their Canvas account. This PR adds an option in the "Sync From Salesforce" page to have newly created users who were never synced before receive an account creation sign-up email. The plan is to expose this link from Salesforce so that if the staff member has a Platform account with the `:CanSyncFromSalesforce` role (or admin), they'll be able to manually run sync and have these emails go out.

Task: https://app.asana.com/0/1174274412967132/1200239858551412

Follow-up task to write specs: https://app.asana.com/0/1174274412967132/1200239858551419

### Motivation and Context
The last minute scramble to close out recruitment has quite a few last minute Participants added and we need to make sure they can login. Note that we can still enhance Campaign Monitor journeys to send the Welcome email to new Participants as well after the initial blast, but this allows regional staff who are working live with folks to get them Canvas access by confirming them and running Sync From Salesforce (e.g. someone just showed up at LL1 who is not in our recruitment pipeline).

This user story is now more important b/c of the one-time token's we're switching sign-up to use, meaning the sign-up links won't be generated until Sync From Salesforce runs. See https://app.asana.com/0/1174274412967132/1200147504835146

#### Pre-populated program and email when opened from Salesforce with new "send sign-up emails" option
image

#### Confirmation page listing users who will get emails
image

#### Message about no users getting emails on confirmation page if we have Users for all Participants
image

#### Actual sign-up email hitting Inbox
image

#### Sample Welcome email from Campaign Monitor that these users missed
image

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] I have added neccessary labels
- [ ] My PR title follows the convention in the guide
- [ ] I have assigned myself to the PR
- [ ] I have requested reviews from team members
- [ ] I have created a task for reviewers on PM tool
- [ ] All new and existing tests passed.

### CODE REVIEW

File: ' [hunk 1]
```diff
@@ -0,0 +1,26 @@
+<% content_for(:headline) do %>
```
Comment (sadleb) [inline ':1]: Ignore this file. Somehow got a file name `'` in my root and accidentally added it. Trying to delete.

File: app/controllers/salesforce_controller.rb [hunk 2]
```diff
@@ -138,4 +148,9 @@ def change_canvas_email()
     raise
   end
 
+  def get_participants_never_synced_before
+    participants = SalesforceAPI.client.find_participants_by(program_id: params[:program_id])
```
Comment (sadleb) [inline app/controllers/salesforce_controller.rb:152]: TODO: also filter by Enrolled status.

File: app/controllers/salesforce_controller.rb [hunk 3]
```diff
@@ -138,4 +148,9 @@ def change_canvas_email()
...
+    participants.select { |p| User.find_by(salesforce_id: p.contact_id).nil? }.compact
+  end
+
```
Comment (sadleb) [inline app/controllers/salesforce_controller.rb:155]: TODO: also filter by Enrolled status.

File: app/models/user.rb [hunk 4]
```diff
@@ -111,6 +112,21 @@ def canvas_login_email
...
+
+    # TODO: use token instead of SF ID.
+    # https://app.asana.com/0/1174274412967132/1200147504835146/f
+    # Make sure you update the link in app/controllers/users/passwords_controller.rb too
+    sign_up_url = new_user_registration_url(u: salesforce_id, protocol: 'https')
+
+    SendSignUpEmailMailer.with(first_name: first_name, sign_up_url: sign_up_url)
+      .sign_up_email.deliver_now
+  end
+
```
Comment (sadleb) [inline app/models/user.rb:129]: I didn't end up consolidating the path in passwords_controller here, b/c this is a full URL.

File: app/policies/salesforce_authorization_policy.rb [hunk 5]
```diff
@@ -4,6 +4,10 @@ def init_sync_from_salesforce_program?
     user.can_sync_from_salesforce?
   end
 
+  def confirm_send_sign_up_emails?
```
Comment (rshipp) [inline app/policies/salesforce_authorization_policy.rb:7]: i'm having trouble parsing what this function does from the name/usage, maybe add a comment?

File: app/services/sync_portal_enrollments_for_program.rb [hunk 6]
```diff
@@ -104,9 +105,22 @@ def find_or_create_user!(sf_participant)
...
       )
+
       # Don't send confirmation email yet; we do that at sign_up time instead.
+      #
+      # Note that there are 3 scenarios where the user gets this email to confirm
+      # and activate their account:
+      # 1) After they get a sign_up link, create their password, and register their account
+      # 2) After they use the password reset link with an account that has never been registered,
+      #    and are sent to create their account instead
+      # 3) After a staff member changes their login email (from Salesforce)
```
Comment (rshipp) [inline app/services/sync_portal_enrollments_for_program.rb:116]: also if they manually request the confirmation email to be sent (there are buttons to request a new email: after attempting to log in with correct credentials for an unconfirmed account, and after successfully registering a new account)

File: app/policies/salesforce_authorization_policy.rb (hunk 5 above)
Comment (rshipp) [reply app/policies/salesforce_authorization_policy.rb:7]: ohhhh this is the policy, nevermind, I totally thought this was the model.

**Review by rshipp — APPROVED** [review]
this seems good, I don't see anything we'd have to make major changes to

File: app/services/sync_portal_enrollments_for_program.rb [hunk 7]
```diff
@@ -104,9 +105,24 @@ def find_or_create_user!(sf_participant)
...
+
       # Don't send confirmation email yet; we do that at sign_up time instead.
+      #
+      # Note that there are 4 scenarios where the user gets this email to confirm
+      # and activate their account:
+      # 1) After they get a sign_up link, create their password, and register their account
+      # 2) After they use the password reset link with an account that has never been registered,
+      #    and are sent to create their account instead
+      # 3) After manually requesting a new confirmation email when trying to log in with
+      #    valid credentials for an confirmed account.
```
Comment (sadleb) [inline app/services/sync_portal_enrollments_for_program.rb:117]: TODO: typo

**Pull request merged**